sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp

```diff
@@ -0,0 +1,100 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <CL/sycl/detail/group_sort_impl.hpp>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace ext {
+namespace oneapi {
+namespace experimental {
+
+// ---- group helpers
+template <typename Group, std::size_t Extent> class group_with_scratchpad {
+  Group g;
+  sycl::span<std::uint8_t, Extent> scratch;
+
+public:
+  group_with_scratchpad(Group g_, sycl::span<std::uint8_t, Extent> scratch_)
+      : g(g_), scratch(scratch_) {}
+  Group get_group() const { return g; }
+  sycl::span<std::uint8_t, Extent> get_memory() const { return scratch; }
+};
+
+// ---- sorters
+template <typename Compare = std::less<>> class default_sorter {
+  Compare comp;
+  std::uint8_t *scratch;
+  std::size_t scratch_size;
+
+public:
+  template <std::size_t Extent>
+  default_sorter(sycl::span<std::uint8_t, Extent> scratch_,
+                 Compare comp_ = Compare())
+      : comp(comp_), scratch(scratch_.data()), scratch_size(scratch_.size()) {}
+
+  template <typename Group, typename Ptr>
+  void operator()(Group g, Ptr first, Ptr last) {
+#ifdef __SYCL_DEVICE_ONLY__
+    using T = typename sycl::detail::GetValueType<Ptr>::type;
+    if (scratch_size >= memory_required<T>(Group::fence_scope, last - first))
+      sycl::detail::merge_sort(g, first, last - first, comp, scratch);
+      // TODO: it's better to add else branch
+#else
+    (void)g;
+    (void)first;
+    (void)last;
+    throw runtime_error(
+        "default_sorter constructor is not supported on host device.",
+        PI_INVALID_DEVICE);
+#endif
+  }
+
+  template <typename Group, typename T> T operator()(Group g, T val) {
+#ifdef __SYCL_DEVICE_ONLY__
+    auto range_size = g.get_local_range().size();
+    if (scratch_size >= memory_required<T>(Group::fence_scope, range_size)) {
+      auto id = sycl::detail::Builder::getNDItem<Group::dimensions>();
+      uint32_t local_id = id.get_local_id();
+      T *temp = reinterpret_cast<T *>(scratch);
+      temp[local_id] = val;
+      sycl::detail::merge_sort(g, temp, range_size, comp,
+                               scratch + range_size * sizeof(T));
+      val = temp[local_id];
+    }
+    // TODO: it's better to add else branch
+#else
+    (void)g;
+    (void)val;
+    throw runtime_error(
+        "default_sorter operator() is not supported on host device.",
+        PI_INVALID_DEVICE);
+#endif
+    return val;
+  }
+
+  template <typename T>
+  static constexpr std::size_t memory_required(sycl::memory_scope scope,
+                                               std::size_t range_size) {
+    return range_size * sizeof(T);
+  }
+
+  template <typename T, int dim = 1>
+  static constexpr std::size_t memory_required(sycl::memory_scope scope,
+                                               sycl::range<dim> r) {
+    return 2 * r.size() * sizeof(T);
```

Comment:
why the `* 2`  ?  Shouldn't `r.size() * sizeof(T)` be sufficient ?  And, if not, why does the `memory_required` variant that takes a `size_t` not also multiply by 2 ?

Reply:
Sure. It's mentioned in the description of the PR. (https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc)

Regarding your comment about tests. It's also mentioned in the description :) See here: https://github.com/intel/llvm-test-suite/pull/438